### PR TITLE
refactor(sdk): replace get_printer(jupyter) by new_printer()

### PR DESCRIPTION
**Pull request by timoffex**
Description
---
`get_printer()` was always called with the result of `settings._jupyter`, which was always computed from `ipython.in_jupyter()`. There's no reason for this to be configurable: if we're running in Jupyter, we should use Jupyter-friendly output.

Because of this parameter, wandb_init.py was deferring printer creation until a Settings object was available and sprinkling `assert self.printer` in various places since `self.printer`'s type had to be `Printer | None`. Now, the printer can be created in the initializer and is guaranteed to be non-None.

Replaces `get_printer()` with `new_printer()` which does not accept any argument.

### CODE REVIEW

File: wandb/sdk/lib/printer.py [hunk 1]
```diff
@@ -64,6 +64,14 @@
 """Color from the 256-color palette for the progress symbol."""
 
 
+def new_printer() -> Printer:
+    """Returns a new printer based on the environment we're in."""
+    if ipython.in_jupyter():
+        return _PrinterJupyter()
+    else:
+        return _PrinterTerm()
```
Comment (kptkin) [inline wandb/sdk/lib/printer.py:72]:
```suggestion
     return _PrinterTerm()
```
nit

**Review by kptkin — APPROVED** [review]

File: wandb/sdk/lib/printer.py (hunk 1 above)
Comment (timoffex) [reply wandb/sdk/lib/printer.py:72]: can do, but why?

Comment (kptkin) [reply wandb/sdk/lib/printer.py:72]: just i think it is better style in python if you have already a return you don't need the else for the other return

Comment (kptkin) [reply wandb/sdk/lib/printer.py:72]: ok, tried to google it, because i know that would be the case in python but looks like the opinions are varied

Comment (timoffex) [reply wandb/sdk/lib/printer.py:72]: they have slightly different meanings, by convention

```python
# If statements are edge cases, final return is the normal path:
if cond:
    return edge_case
return default


# All cases are equal, no case is preferred:
if cond:
    return cond_result
else:
    return other_result
```

Comment (kptkin) [reply wandb/sdk/lib/printer.py:72]: yeah, i just know the first is more idiomatic to python, but i don't have any preference, so if you feel the current is better that's ok :)

**Pull request merged**